ADAL/tests/unit/ios/ADEnrollmentGatewayTests.m

```diff
@@ -0,0 +1,182 @@
+    XCTAssertNil([ADEnrollmentGateway enrollmentIdForUserObjectId:@"d3444455-mike-4271-b6ea-e499cc0cab46" tenantId:@"fda5d5d9-17c3-4c29-9cf9-a27c3d3f03e1" error:NULL]);
+    XCTAssertNil([ADEnrollmentGateway enrollmentIdForUniqueAccountId:@"60406d5d-mike-41e1-aa70-e97501076a22" error:NULL]);
+
+    // dictionary of dictionaries
+    [ADEnrollmentGateway setEnrollmentIdsWithJsonBlob:[NSString stringWithFormat:@"{\"enrollment_ids\":{\"enrollment_ids\":{\"enrollment_ids\":\"enrollment_ids\"}}}"]];
+
+    XCTAssertNil([ADEnrollmentGateway enrollmentIdForUserId:@"mike@contoso.com" error:NULL]);
+    XCTAssertNil([ADEnrollmentGateway enrollmentIdForUserObjectId:@"d3444455-mike-4271-b6ea-e499cc0cab46" tenantId:@"fda5d5d9-17c3-4c29-9cf9-a27c3d3f03e1" error:NULL]);
+    XCTAssertNil([ADEnrollmentGateway enrollmentIdForUniqueAccountId:@"60406d5d-mike-41e1-aa70-e97501076a22" error:NULL]);
+
+    // top level is array instead of dictionary
+    [ADEnrollmentGateway setEnrollmentIdsWithJsonBlob:[NSString stringWithFormat:@"[{\"user_id\":\"mike@contoso.com\"},{\"unique_account_id\":\"60406d5d-mike-41e1-aa70-e97501076a22\"}]"]];
+
+    XCTAssertNil([ADEnrollmentGateway enrollmentIdForUserId:@"mike@contoso.com" error:NULL]);
+    XCTAssertNil([ADEnrollmentGateway enrollmentIdForUserObjectId:@"d3444455-mike-4271-b6ea-e499cc0cab46" tenantId:@"fda5d5d9-17c3-4c29-9cf9-a27c3d3f03e1" error:NULL]);
+    XCTAssertNil([ADEnrollmentGateway enrollmentIdForUniqueAccountId:@"60406d5d-mike-41e1-aa70-e97501076a22" error:NULL]);
+
+    // enrollmentId missing
+    [ADEnrollmentGateway setEnrollmentIdsWithJsonBlob:[NSString stringWithFormat:
+                                                       @"{\"enrollment_ids\": [\n"
+                                                       "{\n"
+                                                       "\"tid\" : \"fda5d5d9-17c3-4c29-9cf9-a27c3d3f03e1\",\n"
+                                                       "\"oid\" : \"d3444455-mike-4271-b6ea-e499cc0cab46\",\n"
+                                                       "\"unique_account_id\" : \"60406d5d-mike-41e1-aa70-e97501076a22\",\n"
+                                                       "\"user_id\" : \"mike@contoso.com\",\n"
+                                                       "}}"]];
+
+    XCTAssertNil([ADEnrollmentGateway enrollmentIdForUserId:@"mike@contoso.com" error:NULL]);
+    XCTAssertNil([ADEnrollmentGateway enrollmentIdForUserObjectId:@"d3444455-mike-4271-b6ea-e499cc0cab46" tenantId:@"fda5d5d9-17c3-4c29-9cf9-a27c3d3f03e1" error:NULL]);
+    XCTAssertNil([ADEnrollmentGateway enrollmentIdForUniqueAccountId:@"60406d5d-mike-41e1-aa70-e97501076a22" error:NULL]);
+
+}
+
+- (void) testEnrollmentIDErrorsReturned
+{
+    // dictionary of dictionaries
+    [ADEnrollmentGateway setEnrollmentIdsWithJsonBlob:[NSString stringWithFormat:@"{\"enrollment_ids\":{\"enrollment_ids\":{\"enrollment_ids\":\"enrollment_ids\"}}}"]];
+
+    // Valid JSON does not return error
```

Comment:
In this case, I could also generate a new error object and populate it with a message stating that the JSON doesn't conform to the expected structure, but I didn't see any existing examples of this pattern, so I left it as log at WARN level, let me know if I should change this.

Reply:
here is an example:
https://github.com/AzureAD/azure-activedirectory-library-for-objc/blob/release/2.6.4/ADAL/src/request/ADAuthenticationRequest%2BAcquireToken.m#L207

We could add an addition error code, say `AD_ERROR_ENROLLMENT_IDS_BAD_FORMAT` :)